### PR TITLE
Fix arm64 wrong multireg copies from reg to small type field in memory

### LINKED ISSUE #46240: wrong multireg copies from reg to small type field in memory
Add a repro for the problem that was pointed in https://github.com/dotnet/runtime/pull/46176#discussion_r545493173

```
using System;
using System.Diagnostics;
using System.Runtime.CompilerServices;
using System.Numerics;

public struct S
{
    public object o;
    public byte b;
}

public class UnitTest1
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    public static void TestS(S s)
    {
        return;
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    public static S GetS()
    {
        return new S();
    }



    [MethodImpl(MethodImplOptions.NoInlining)]
    public static int TestS()
    {

        S s1 = GetS();
        S s2 = GetS();
        S s3 = GetS();
        S s4 = GetS();
        S s5 = GetS();
        S s6 = GetS();
        S s7 = GetS();
        S s8 = GetS();
        S s9 = GetS();
        S s10 = GetS();

        TestS(s10);
        TestS(s9);
        TestS(s8);
        TestS(s7);
        TestS(s6);
        TestS(s5);
        TestS(s4);
        TestS(s3);
        TestS(s2);
        TestS(s1);

        return 1;

    }

    public static int Main(string[] args)
    {
        TestS();
        return 100;
    }
}
```

It has regressed after https://github.com/dotnet/runtime/pull/46176 + https://github.com/dotnet/runtime/pull/45625

Fails with
```
Assert failure(PID 20984 [0x000051f8], Thread: 34716 [0x879c]): Assertion failed 'isValidGeneralDatasize(size)' in 'UnitTest1:TestS():int' during 'Generate code' (IL size 102)

    File: D:\Sergey\git\runtime\src\coreclr\jit\emitarm64.cpp Line: 7689
```
because we use different size for `ins_Store(8 byte in this case)` and `emitTypeSize(1 byte in this case)`.

repro steps:
```
complus_altjit=*
complus_altjitname=clrjit_win_arm64_x64.dll
complus_SIMD16ByteOnly=1
complus_jitdump=UnitTest1:TestS():int
windows.x64.Checked\Tests\Core_Root\CoreRun.exe test.dll
```

### PR DESCRIPTION
**Pull request by sandreenko**
Closes  #46240

### CODE REVIEW

**Comment by sandreenko** [comment]
Can anybody approve this?

**Review by AndyAyersMS — APPROVED** [review]

**Comment by sandreenko** [comment]
@AndyAyersMS I've changed the fix, please take another look.  cc  @dotnet/jit-contrib , @jkoritzinsky

File: src/coreclr/jit/codegencommon.cpp [hunk 1]
```diff
@@ -11855,15 +11856,34 @@ void CodeGen::genMultiRegStoreToLocal(GenTreeLclVar* lclNode)
...
+#ifdef TARGET_64BIT
+            assert(offset <= varDsc->lvSize());
+#else  // !TARGET_64BIT
+            if (varTypeIsStruct(varDsc))
+            {
+                assert(offset <= varDsc->lvSize());
+            }
+            else
+            {
+                assert((varDsc->TypeGet() == TYP_LONG));
```
Comment (sandreenko) [inline src/coreclr/jit/codegencommon.cpp:11882]:
```suggestion
                assert(varDsc->TypeGet() == TYP_LONG);
```

File: src/coreclr/jit/instr.cpp [hunk 2]
```diff
@@ -1963,6 +1963,44 @@ instruction CodeGenInterface::ins_Store(var_types dstType, bool aligned /*=false
...
+// Arguments:
+//   srcReg  - the source register for the store
+//   dstType - the destination type
+//   aligned - whether the destination is properly aligned if dstType is a SIMD type
+//
+// Return Value:
+//   the instruction to use
+//
+// Notes:
+//   The fuctiuon currently does not expect float srcReg with integral dstType and will assert on such cases.
```
Comment (AndyAyersMS) [inline src/coreclr/jit/instr.cpp:1978]:
```suggestion
//   The function currently does not expect float srcReg with integral dstType and will assert on such cases.
```

File: src/coreclr/jit/instr.cpp [hunk 3]
```diff
@@ -1963,6 +1963,44 @@ instruction CodeGenInterface::ins_Store(var_types dstType, bool aligned /*=false
...
+//
+instruction CodeGenInterface::ins_Store(regNumber srcReg, var_types dstType, bool aligned /*=false*/)
```
Comment (AndyAyersMS) [inline src/coreclr/jit/instr.cpp:1980]: Would prefer not to see overloads like this with optional parameters. Can we rename one or the other?

Comment (AndyAyersMS) [reply src/coreclr/jit/instr.cpp:1980]: Also not clear to me why we don't just force all stores to use the version that does more checking...

Comment (sandreenko) [reply src/coreclr/jit/instr.cpp:1980]: Makes sense, I am thinking how to rename one of them now.

> Also not clear to me why we don't just force all stores to use the version that does more checking...

It is possible but requires more refactoring changes, for example, `emitInsStoreLcl` and `emitInsStoreInd` should just receive `GenTree*` and calculate `instruction ins, emitAttr attr` inside. I will try to find time for this change later as a separate PR.

**Review by AndyAyersMS — APPROVED** [review]
LGTM

**Comment by danmosemsft** [comment]
did this fix https://github.com/dotnet/runtime/issues/46278 ?

**Comment by sandreenko** [comment]
> did this fix #46278 ?

Yes, thanks for pointing this, I forgot to link that issue.

**Pull request merged**